### ISSUE TITLE
fail to narrow enum typed variable when matching a tuple of enum values using `in`

### ISSUE THREAD

**Issue by programmerjake** [microsoft/pyright#4102]
**Describe the bug**

enum typed variables aren't narrowed when matching using `in`.

**To Reproduce**

type check the following code:
```python
import enum
from typing import NoReturn

class MyEnum(enum.Enum):
    A = enum.auto()
    B = enum.auto()
    C = enum.auto()

    def is_a_or_b(self) -> bool:
        if self in (MyEnum.A, MyEnum.B):
            return True
        if self is MyEnum.C:
            return False
        # ensure we explicitly matched all enum values
        assert_never(self)  # incorrectly produces a type error here

def assert_never(x: NoReturn) -> NoReturn:
    raise ValueError
```

typechecker output:
```
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/test_prj/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.277
/tmp/test_prj/__init__.py
  /tmp/test_prj/__init__.py:15:22 - error: Argument of type "Literal[MyEnum.A, MyEnum.B]" cannot be assigned to parameter "x" of type "NoReturn" in function "assert_never"
    Type "Literal[MyEnum.A, MyEnum.B]" cannot be assigned to type "NoReturn"
      Type "Literal[MyEnum.A]" cannot be assigned to type "NoReturn" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.522sec
```

**Expected behavior**
No error to be produced, since `is_a_or_b` did match against every enum value.

**VS Code extension or command-line**
command line pyright 1.1.277

**Comment by erictraut** [comment]
The `x in y` type narrowing pattern currently does not expand enums into their constituent parts in the negative ("else") narrowing case. I'll note that mypy doesn't do this either.

I think it's reasonable (for consistency) to add this capability.

In the meantime, you can work around the issue by doing one of the following:
1. Swap the order of the two `if` statements in your code above. By moving the `x is y` conditional first, the narrowed type is expanded from the full enum to a union of its constituent elements in the negative ("else") case. The subsequent `x in y` narrowing logic is then able to handle the further narrowing.
2. Avoid the use of `x in y` pattern and replace it with `x is y or x is z`.

**Comment by erictraut** [comment]
This is included in pyright 1.1.278, which I just published. It will also be included in a future release of pylance.

**Comment by programmerjake** [comment]
Thanks for quickly addressing this!
